### PR TITLE
[Brave News]: Implement FeedV2 sidebar, for changing feed view

### PR DESCRIPTION
**Pull request by fallaciousreasoning**
Resolves https://github.com/brave/brave-browser/issues/33672

Sidebar (all):
image

[Screencast from 2023-10-25 10-38-13.webm](https://github.com/brave/brave-core/assets/7678024/28da9018-dda0-457f-b5be-b015b5c60d96)

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/brave_news/browser/feed_v2_builder.cc [hunk 1]
```diff
@@ -434,7 +518,27 @@ void FeedV2Builder::FetchFeed() {
 void FeedV2Builder::OnFetchedFeed(FeedItems items, ETags tags) {
   DVLOG(1) << __FUNCTION__;
 
-  raw_feed_items_ = std::move(items);
+  raw_feed_items_.clear();
+  raw_feed_items_.reserve(items.size());
+
+  // We can get duplicate articles from different regions.
+  base::flat_set<GURL> seen;
```
Comment (fallaciousreasoning) [inline components/brave_news/browser/feed_v2_builder.cc:525]: This should be pushed down into the `FeedFetcher`

File: components/brave_news/browser/resources/Sidebar.tsx [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+const Heading = styled.h3`
+  font: ${font.primary.default.semibold};
+  color: var(--bn-card-color-0.75);
+  margin: 0;
+
+  padding-left: ${PAD_LEFT};
+`
+
+const CustomButton = styled.button <{ selected?: boolean, faint?: boolean, bold?: boolean }>`
```
Comment (fallaciousreasoning) [inline components/brave_news/browser/resources/Sidebar.tsx:32]: cc @petemill this makes me sad

File: components/brave_news/browser/resources/Sidebar.tsx [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+  display: block;
+  outline: none;
+  border-radius: ${radius.m};
+  border: none;
+  background: none;
+  text-align: left;
+  width: 100%;
+
+  color: ${p => `rgba(var(--bn-text-base), ${p.faint ? 0.25 : 1})`};
```
Comment (fallaciousreasoning) [inline components/brave_news/browser/resources/Sidebar.tsx:43]: cc @petemill I haven't worked out a complete theming story yet - everything is very custom.

File: components/brave_news/browser/resources/Sidebar.tsx [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+  }
+
+  ${p => p.selected && css`background: rgba(255,255,255,0.3);`}
+
+  & > * {
+    text-align: left;
+  }
+`
+const Section = styled.details`
+  & summary {
```
Comment (fallaciousreasoning) [inline components/brave_news/browser/resources/Sidebar.tsx:62]: Sadly I can't use the Leo `Collapse` here :cry:

File: components/brave_news/browser/resources/Sidebar.tsx [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+    --leo-icon-size: 18px;
+
+    color: ${color.text.primary};
```
Comment (fallaciousreasoning) [inline components/brave_news/browser/resources/Sidebar.tsx:65]: Note to self: Recheck colors - any Leo color usage is **probably** wrong

File: components/brave_news/browser/resources/context.tsx [hunk 6]
```diff
@@ -9,8 +9,11 @@ import {
 import * as React from 'react'
 import usePromise from '../../../brave_new_tab_ui/hooks/usePromise'
 
+export type FeedType = 'all' | `publishers/${string}` | `channels/${string}`
```
Comment (fallaciousreasoning) [inline components/brave_news/browser/resources/context.tsx:12]: This is going to conflict with https://github.com/brave/brave-core/pull/20452

File: components/brave_news/browser/brave_news_controller.cc [hunk 7]
```diff
@@ -152,6 +152,22 @@ void BraveNewsController::ClearHistory() {
   // feed cache somewhere.
 }
 
+bool BraveNewsController::MaybeInitFeedV2() {
```
Comment (sangwoo108) [inline components/brave_news/browser/brave_news_controller.cc:155]: Nice!

File: components/brave_news/browser/resources/context.tsx (hunk 6 above)
Comment (sangwoo108) [reply components/brave_news/browser/resources/context.tsx:12]: It's interesting that we can use template literal like this 👍

**Review by sangwoo108 — APPROVED** [review]
LGTM 😄

File: components/brave_news/browser/resources/context.tsx (hunk 6 above)
Comment (fallaciousreasoning) [reply components/brave_news/browser/resources/context.tsx:12]: https://www.typescriptlang.org/docs/handbook/2/template-literal-types.html

They're really handy, especially when you have combinations of strings that'd be painful to write out manually:

```
type Vertical = 'top' | 'bottom' | 'center'
type Horizontal = 'left' | 'right' | 'center'
type Align = `${Vertical}-${Horizontal}` | `center`

**Pull request merged**